api/network/domains_api.go

```diff
@@ -0,0 +1,247 @@
+package network
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/ingrammicro/cio/api/types"
+	"github.com/ingrammicro/cio/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+// DomainService manages DNS domain and record operations
+type DomainService struct {
+	concertoService utils.ConcertoService
+}
+
+// NewDomainService returns a Concerto Domain service
+func NewDomainService(concertoService utils.ConcertoService) (*DomainService, error) {
+	if concertoService == nil {
+		return nil, fmt.Errorf("must initialize ConcertoService before using it")
+	}
+
+	return &DomainService{
+		concertoService: concertoService,
+	}, nil
+}
+
+// ListDomains returns the list of domains as an array of Domain
+func (ds *DomainService) ListDomains() (domains []*types.Domain, err error) {
+	log.Debug("ListDomains")
+
+	data, status, err := ds.concertoService.Get("/network/dns/domains")
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err = utils.CheckStandardStatus(status, data); err != nil {
+		return nil, err
+	}
+
+	if err = json.Unmarshal(data, &domains); err != nil {
+		return nil, err
+	}
+
+	return domains, nil
+}
+
+// GetDomain returns a domain by its ID
+func (ds *DomainService) GetDomain(domainID string) (domain *types.Domain, err error) {
+	log.Debug("GetDomain")
+
+	data, status, err := ds.concertoService.Get(fmt.Sprintf("/network/dns/domains/%s", domainID))
+	if err != nil {
+		return nil, err
+	}
+
+	if err = utils.CheckStandardStatus(status, data); err != nil {
+		return nil, err
+	}
+
+	if err = json.Unmarshal(data, &domain); err != nil {
+		return nil, err
+	}
+
+	return domain, nil
+}
+
+// CreateDomain creates a domain
+func (ds *DomainService) CreateDomain(domainParams *map[string]interface{}) (domain *types.Domain, err error) {
+	log.Debug("CreateDomain")
+
+	data, status, err := ds.concertoService.Post("/network/dns/domains", domainParams)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = utils.CheckStandardStatus(status, data); err != nil {
+		return nil, err
+	}
+
+	if err = json.Unmarshal(data, &domain); err != nil {
+		return nil, err
+	}
+
+	return domain, nil
+}
+
+// DeleteDomain deletes a domain by its ID
+func (ds *DomainService) DeleteDomain(domainID string) (domain *types.Domain, err error) {
+	log.Debug("DeleteDomain")
+
+	data, status, err := ds.concertoService.Delete(fmt.Sprintf("/network/dns/domains/%s", domainID))
+	if err != nil {
+		return nil, err
+	}
+
+	if err = utils.CheckStandardStatus(status, data); err != nil {
+		return nil, err
+	}
+
+	if err = json.Unmarshal(data, &domain); err != nil {
+		return nil, err
+	}
+
+	return domain, nil
+}
+
+// RetryDomain retries a domain by its ID
+func (ds *DomainService) RetryDomain(domainID string, domainParams *map[string]interface{}) (domain *types.Domain, err error) {
```

Comment:
I don't think it makes sense to have a domainParams for this method: retry is an action on a domain, the same as boot is for a server...

Reply:
Yes, it is not really necessary; since no parameters are required for these action types.
But it is done this way as a procedure or pattern, it is always used this way for PUT methods

Reply:
Done